### PR TITLE
Various Karl-related tweaks + `[Immutable]Optional<>` comparisons.

### PR DESCRIPTION
**Pull request by mzhurovich**
Sorry for some mess, but I hope this PR is good for per-commit review. Thanks!

### CODE REVIEW

File: Karl/integrations/nodejs/test.js [hunk 1]
```diff
@@ -23,17 +23,20 @@ SOFTWARE.
...
 
 var assert = require('assert');
+var fs = require('fs');
 var request = require('request');
 var Claire = require('./claire.js');
 
 const testKarlKeepalivesPort = 30001;
 const testKarlFleetViewPort = 30999;
 
+const currentBuild = JSON.parse(fs.readFileSync('current_build.json'));
```
Comment (dkorolev) [inline Karl/integrations/nodejs/test.js]: `console.error` and `process.exit(-1)` on failure?

File: Karl/integrations/nodejs/test.js [hunk 2]
```diff
@@ -105,18 +107,10 @@ describe('Claire JS client test.', function () {
           assert.equal(claire.dependencies[0].ip, '127.0.0.1');
           assert.equal(claire.dependencies[0].port, 8283);
           assert.equal(claire.dependencies[0].prefix, '/');
-          completed = true;
           done();
         });
       });
     });
-
```
Comment (dkorolev) [inline Karl/integrations/nodejs/test.js]: :+1:

File: Karl/karl.h [hunk 3]
```diff
@@ -893,7 +893,7 @@ class GenericKarl final : private KarlStorage<STORAGE_TYPE>,
                            blob.build_time_epoch_microseconds = info.build_time_epoch_microseconds;
                            blob.git_commit = info.git_commit_hash;
                            blob.git_branch = info.git_branch;
-                           blob.git_dirty = !info.git_dirty_files.empty();
+                           blob.git_dirty = Exists(info.git_dirty_files) && !Value(info.git_dirty_files).empty();
```
Comment (dkorolev) [inline Karl/karl.h]: :+1:

File: Karl/render.h [hunk 4]
```diff
@@ -91,7 +91,7 @@ inline graphviz::DiGraph Render(const current::karl::GenericKarlStatus<INNER_STA
       const std::string& codename = e.first;
       const auto& service = e.second;
       std::ostringstream os;
-      os << "<TABLE CELLBORDER='0'>";
+      os << "<TABLE CELLBORDER='0' CELLSPACING='5'>";
```
Comment (dkorolev) [inline Karl/render.h]: R U sure it won't break our prod layouts?

File: Karl/render.h [hunk 5]
```diff
@@ -137,54 +137,53 @@ inline graphviz::DiGraph Render(const current::karl::GenericKarlStatus<INNER_STA
 
       // Second section, build info.
       {
-        auto text = service.git_commit;
-        if (text.empty()) {
-          text = "build";
-        } else if (text.length() > 6) {
-          text = text.substr(0, 6);
-        }
+        const std::string commit_text = Exists(service.git_commit) ? Value(service.git_commit).substr(0, 6) : "build";
```
Comment (dkorolev) [inline Karl/render.h]: `"build"` reads confusing. The message is that it's out-of-git, manual, build, right?

File: Karl/schema_karl.h [hunk 6]
```diff
@@ -303,19 +303,19 @@ CURRENT_STRUCT_T(ServiceToReport) {
   CURRENT_FIELD_DESCRIPTION(build_time,
                             "An exempt of this service's build information, to help visually locating a "
                             "specific, usually recently started, service.");
-  CURRENT_FIELD(build_time_epoch_microseconds, std::chrono::microseconds);
+  CURRENT_FIELD(build_time_epoch_microseconds, std::chrono::microseconds, std::chrono::microseconds(0));
```
Comment (dkorolev) [inline Karl/schema_karl.h]: :+1:

File: TypeSystem/optional.h [hunk 7]
```diff
@@ -65,14 +66,16 @@ class ImmutableOptional<T, std::enable_if_t<std::is_pod<T>::value>> final {
 
   bool ExistsImpl() const { return exists_; }
 
-  T ValueImpl() const {
+  T operator*() const {
```
Comment (dkorolev) [inline TypeSystem/optional.h]: Are you sure it's a good idea? Looks confusingly scary to me TBH, `Exists(x)`+`Value(x)` уже такие тёплые ламповые, why allow alternate syntax?

File: TypeSystem/optional.h [hunk 8]
```diff
@@ -99,19 +102,114 @@ class ImmutableOptional<T, std::enable_if_t<!std::is_pod<T>::value>> final {
...
+
  private:
   std::unique_ptr<T> owned_optional_object_;
   const T* optional_object_;
 };
 
+// Compare `ImmutableOptional<T>` with `ImmutableOptional<T>`.
+template <class T>
+bool operator==(const ImmutableOptional<T>& x, const ImmutableOptional<T>& y) {
+  return x.ExistsImpl() != y.ExistsImpl() ? false : x.ExistsImpl() == false ? true : *x == *y;
```
Comment (dkorolev) [inline TypeSystem/optional.h]: `s/ExistsImpl() == false/` -> flip the order? Also, parentheses would help.

File: TypeSystem/optional.h [hunk 9]
```diff
@@ -99,19 +102,114 @@ class ImmutableOptional<T, std::enable_if_t<!std::is_pod<T>::value>> final {
...
+}
+
+template <class T>
+bool operator!=(const ImmutableOptional<T>& x, const ImmutableOptional<T>& y) {
+  return !(x == y);
+}
+
+template <class T>
+bool operator<(const ImmutableOptional<T>& x, const ImmutableOptional<T>& y) {
+  return !y.ExistsImpl() ? false : !x.ExistsImpl() ? true : *x < *y;
```
Comment (dkorolev) [inline TypeSystem/optional.h]: Parentheses would help here too.

File: TypeSystem/optional.h [hunk 10]
```diff
@@ -99,19 +102,114 @@ class ImmutableOptional<T, std::enable_if_t<!std::is_pod<T>::value>> final {
...
+bool operator<=(const ImmutableOptional<T>& x, const ImmutableOptional<T>& y) {
+  return !(y < x);
+}
+
+template <class T>
+bool operator>=(const ImmutableOptional<T>& x, const ImmutableOptional<T>& y) {
+  return !(x < y);
+}
+
+// Compare `ImmutableOptional<T>` with `const T&`.
```
Comment (dkorolev) [inline TypeSystem/optional.h]: :+1:

File: TypeSystem/test.cc [hunk 11]
```diff
@@ -619,6 +619,219 @@ TEST(TypeSystemTest, Optional) {
...
+  EXPECT_FALSE(i1 >= 2);
+  EXPECT_TRUE(2 >= i1);
+  EXPECT_FALSE(empty >= 1);
+  EXPECT_TRUE(1 >= empty);
+}
+
+template <template <typename...> class OPTIONAL_TYPE>
+void TestComparisonOperatorsForNonPOD() {
+  OPTIONAL_TYPE<std::string> i1("1"), j1("1"), k2("2");
+  OPTIONAL_TYPE<std::string> empty(nullptr), empty2(nullptr);
```
Comment (dkorolev) [inline TypeSystem/test.cc]: A test for a custom `CURRENT_STRUCT` with its own comparison operators, maybe?

File: scripts/gen-current-build-json.sh [hunk 12]
```diff
@@ -0,0 +1,44 @@
+#!/bin/bash
+
+# Generates the minimalistic `current_build.json` file without compiler-related fields.
+
+BUILD_DATETIME=$(LC_ALL=en_EN date "+%b %d %Y, %H:%M:%S")
+BUILD_EPOCH_MICROSECONDS="$(date "+%s")000000"
+
+OS_VERSION="$(uname -a)"
+if [[ $? -ne 0 ]]; then
+	exit 1
```
Comment (dkorolev) [inline scripts/gen-current-build-json.sh]: Wait, are these tabs?

File: scripts/gen-current-build-json.sh [hunk 13]
```diff
@@ -0,0 +1,44 @@
+#!/bin/bash
+
+# Generates the minimalistic `current_build.json` file without compiler-related fields.
+
+BUILD_DATETIME=$(LC_ALL=en_EN date "+%b %d %Y, %H:%M:%S")
+BUILD_EPOCH_MICROSECONDS="$(date "+%s")000000"
```
Comment (dkorolev) [inline scripts/gen-current-build-json.sh]: I know it's OK, but `s/date "+%s"/date '+%s'/` would make me sleep better.

File: scripts/gen-current-build-json.sh [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+	OPTIONAL_GIT_FIELDS+="  \"git_branch\": \"$GIT_BRANCH\","
+fi
+
+cat >$1 << EOF
+{
+  "build_time": "$BUILD_DATETIME",
+  "build_dir": "$PWD",
+  "build_user": "$(whoami)",
+  "build_time_epoch_microseconds": $BUILD_EPOCH_MICROSECONDS,
+  $OPTIONAL_GIT_FIELDS
```
Comment (dkorolev) [inline scripts/gen-current-build-json.sh]: :+1:

File: scripts/gen-current-build.sh [hunk 15]
```diff
@@ -169,23 +170,23 @@ CURRENT_STRUCT(BuildInfo) {
...
-    std::string,
-    std::string,
-    std::string>
+    Optional<std::string>,
+    Optional<std::vector<std::string>>,
+    Optional<std::string>,
+    Optional<std::string>,
+    Optional<std::string>,
+    Optional<std::string>,
+    Optional<std::string>>
```
Comment (dkorolev) [inline scripts/gen-current-build.sh]: Sorry about this.

File: Karl/render.h (hunk 4 above)
Comment (mzhurovich) [inline Karl/render.h]: Cellspacing? How?

Comment (dkorolev) [inline Karl/render.h]: Don't shoot me, I'm not a frontend guy.

File: Karl/render.h (hunk 5 above)
Comment (mzhurovich) [inline Karl/render.h]: Yep. I kept the original string. Btw, this text is a link to `./build/codename`. What text could be better in your opinion?

Comment (dkorolev) [inline Karl/render.h]: Best I could think of is "not git tracking".

File: TypeSystem/test.cc (hunk 11 above)
Comment (mzhurovich) [inline TypeSystem/test.cc]: Not now :(

File: scripts/gen-current-build-json.sh (hunk 12 above)
Comment (mzhurovich) [inline scripts/gen-current-build-json.sh]: Yes.

File: scripts/gen-current-build-json.sh (hunk 13 above)
Comment (mzhurovich) [inline scripts/gen-current-build-json.sh]: 👍

File: scripts/gen-current-build-json.sh (hunk 12 above)
Comment (mzhurovich) [inline scripts/gen-current-build-json.sh]: As in `gen-current-build.sh` btw.

File: Karl/render.h (hunk 5 above)
Comment (mzhurovich) [inline Karl/render.h]: Ok. It's right under "commit" header. How about `unknown`?

Comment (mzhurovich) [inline Karl/render.h]: Ok. It's right under "commit" header. How about `unknown`?

**Comment by mzhurovich** [comment]
@dkorolev PTAL https://github.com/C5T/Current/pull/652/commits/145384ebe222d7c0679f977f98f078263188d5a1

**Comment by dkorolev** [comment]
Merging to unblock a related PR. Thanks @mzhurovich @sompylasar!

**Comment by mzhurovich** [comment]
@nyospe FYI I've added comparison operators for Optionals.

**Pull request merged**